File: services/analytics/app/work_list_generator.py
from decimal import Decimal

from app.schemas import WorkItemTemplate, WorkListRequest, WorkListResponse

# Prezzi indicativi €/mq o €/cad — da calibrare per zona e fornitori
BASE_RATES: dict[str, Decimal] = {
    "demolition": Decimal("25"),
    "masonry": Decimal("80"),
    "electrical": Decimal("45"),
    "plumbing": Decimal("50"),
    "hvac": Decimal("35"),
    "windows": Decimal("350"),
    "drywall": Decimal("30"),
    "flooring": Decimal("40"),
    "tiling": Decimal("45"),
    "painting": Decimal("12"),
    "bathroom": Decimal("4500"),
    "kitchen": Decimal("6000"),
    "doors": Decimal("250"),
    "lighting": Decimal("800"),
    "furnishing": Decimal("0"),
    "disposal": Decimal("500"),
    "inspection": Decimal("300"),
}


def generate_work_list(req: WorkListRequest) -> WorkListResponse:
    sqm = req.surface_sqm
    items: list[WorkItemTemplate] = []
    notes: list[str] = []

    if req.condition in ("da_ristrutturare", "pessimo", "ruin"):
        items.append(WorkItemTemplate(
            room="Generale",
            category="demolition",
            description="Demolizioni selettive e rimozione finiture esistenti",
            unit="mq",
            quantity=sqm * Decimal("0.3"),
            unit_price=BASE_RATES["demolition"],
            requires_permit=False,
        ))

    items.extend([
        WorkItemTemplate(
            room="Generale",
            category="masonry",
            description="Opere murarie e rasature",
            unit="mq",
            quantity=sqm,
            unit_price=BASE_RATES["masonry"],
        ),
        WorkItemTemplate(
            room="Generale",
            category="electrical",
            description="Rifacimento impianto eletrico a norma",
            unit="mq",
            quantity=sqm,
            unit_price=BASE_RATES["electrical"],
            requires_permit=True,
        ),
        WorkItemTemplate(
            room="Generale",
            category="plumbing",
            description="Rifacimento impianto idrico-sanitario",
            unit="mq",
            quantity=sqm,
            unit_price=BASE_RATES["plumbing"],
            requires_permit=True,
        ),
        WorkItemTemplate(
            room="Generale",
            category="hvac",
            description="Climatizzazione (split o centralizzato)",
            unit="mq",
            quantity=sqm,
            unit_price=BASE_RATES["hvac"],
        ),
        WorkItemTemplate(
            room="Generale",
            category="windows",
            description="Sostituzione infissi",
            unit="cad",
            quantity=Decimal(max(req.rooms + 1, 4)),
            unit_price=BASE_RATES["windows"],
        ),
        WorkItemTemplate(
            room="Generale",
            category="flooring",
            description="Pavimentazione",
            unit="mq",
            quantity=sqm,
            unit_price=BASE_RATES["flooring"],
        ),
        WorkItemTemplate(
            room="Generale",
            category="painting",
            description="Tinteggiatura pareti e soffitti",
            unit="mq",
            quantity=sqm * Decimal("3.5"),
            unit_price=BASE_RATES["painting"],
        ),
        WorkItemTemplate(
            room="Generale",
            category="disposal",
            description="Smaltimento macerie e trasporto",
            unit="cad",
            quantity=Decimal("1"),
            unit_price=BASE_RATES["disposal"],
        ),
        WorkItemTemplate(
            room="Generale",
            category="inspection",
            description="Collaudo impianti e certificazioni",
            unit="cad",
            quantity=Decimal("1"),
            unit_price=BASE_RATES["inspection"],
            requires_permit=True,
        ),
    ])

    for i in range(req.include_bathrooms):
        items.append(WorkItemTemplate(
            room=f"Bagno {i + 1}",
            category="bathroom",
            description="Rifacimento completo bagno",
            unit="cad",
            quantity=Decimal("1"),
            unit_price=BASE_RATES["bathroom"],
            requires_permit=True,
        ))

    if req.include_kitchen:
        items.append(WorkItemTemplate(
            room="Cucina",
            category="kitchen",
            description="Rifacimento cucina (impianti + mobili base)",
            unit="cad",
            quantity=Decimal("1"),
            unit_price=BASE_RATES["kitchen"],
        ))

    total = sum(
        (item.quantity * item.unit_price for item in items),
        Decimal("0"),
    )

    permit_items = [i for i in items if i.requires_permit]
    if permit_items:
        notes.append(
            f"{len(permit_items)} voci richiedono verifica titoli edilizi/permessi."
        )
    notes.append(
        "Importi indicativi. Validare con preventivi d'impresa prima dell'impegno."
    )
    notes.append(
        "Verificare agevolazioni fiscali applicabili con il commercialista."
    )

    return WorkListResponse(
        items=items,
        total_estimated=total.quantize(Decimal("0.01")),
        notes=notes,
    )

File: apps/web/src/lib/analytics-engine/work-list-generator.ts
const BASE_RATES: Record<string, number> = {
  demolition: 25,
  masonry: 80,
  electrical: 45,
  plumbing: 50,
  hvac: 35,
  windows: 350,
  drywall: 30,
  flooring: 40,
  tiling: 45,
  painting: 12,
  bathroom: 4500,
  kitchen: 6000,
  doors: 250,
  lighting: 800,
  furnishing: 0,
  disposal: 500,
  inspection: 300,
};

export interface WorkItemTemplate {
  room: string;
  category: string;
  description: string;
  unit: string;
  quantity: number;
  unit_price: number;
  requires_permit?: boolean;
}

export interface WorkListRequest {
  surface_sqm: number;
  rooms?: number;
  condition?: string;
  include_kitchen?: boolean;
  include_bathrooms?: number;
}

export interface WorkListResponse {
  items: WorkItemTemplate[];
  total_estimated: number;
  notes: string[];
}

export function generateWorkList(req: WorkListRequest): WorkListResponse {
  const sqm = req.surface_sqm;
  const rooms = req.rooms ?? 3;
  const condition = req.condition ?? "da_ristrutturare";
  const includeBathrooms = req.include_bathrooms ?? 1;
  const items: WorkItemTemplate[] = [];
  const notes: string[] = [];

  if (condition === "da_ristrutturare" || condition === "pessimo" || condition === "ruin") {
    items.push({
      room: "Generale",
      category: "demolition",
      description: "Demolizioni selettive e rimozione finiture esistenti",
      unit: "mq",
      quantity: roundMoney(sqm * 0.3),
      unit_price: BASE_RATES.demolition,
      requires_permit: false,
    });
  }

  items.push(
    {
      room: "Generale",
      category: "masonry",
      description: "Opere murarie e rasature",
      unit: "mq",
      quantity: sqm,
      unit_price: BASE_RATES.masonry,
    },
    {
      room: "Generale",
      category: "electrical",
      description: "Rifacimento impianto elettrico a norma",
      unit: "mq",
      quantity: sqm,
      unit_price: BASE_RATES.electrical,
      requires_permit: true,
    },
    {
      room: "Generale",
      category: "plumbing",
      description: "Rifacimento impianto idrico-sanitario",
      unit: "mq",
      quantity: sqm,
      unit_price: BASE_RATES.plumbing,
      requires_permit: true,
    },
    {
      room: "Generale",
      category: "hvac",
      description: "Climatizzazione (split o centralizzato)",
      unit: "mq",
      quantity: sqm,
      unit_price: BASE_RATES.hvac,
    },
    {
      room: "Generale",
      category: "windows",
      description: "Sostituzione infissi",
      unit: "cad",
      quantity: Math.max(rooms + 1, 4),
      unit_price: BASE_RATES.windows,
    },
    {
      room: "Generale",
      category: "flooring",
      description: "Pavimentazione",
      unit: "mq",
      quantity: sqm,
      unit_price: BASE_RATES.flooring,
    },
    {
      room: "Generale",
      category: "painting",
      description: "Tinteggiatura pareti e soffitti",
      unit: "mq",
      quantity: sqm * 3.5,
      unit_price: BASE_RATES.painting,
    },
    {
      room: "Generale",
      category: "disposal",
      description: "Smaltimento macerie e trasporto",
      unit: "cad",
      quantity: 1,
      unit_price: BASE_RATES.disposal,
    },
    {
      room: "Generale",
      category: "inspection",
      description: "Collaudo impianti e certificazioni",
      unit: "cad",
      quantity: 1,
      unit_price: BASE_RATES.inspection,
      requires_permit: true,
    }
  );

  for (let i = 0; i < includeBathrooms; i += 1) {
    items.push({
      room: `Bagno ${i + 1}`,
      category: "bathroom",
      description: "Rifacimento completo bagno",
      unit: "cad",
      quantity: 1,
      unit_price: BASE_RATES.bathroom,
      requires_permit: true,
    });
  }

  if (req.include_kitchen !== false) {
    items.push({
      room: "Cucina",
      category: "kitchen",
      description: "Rifacimento cucina (impianti + mobili base)",
      unit: "cad",
      quantity: 1,
      unit_price: BASE_RATES.kitchen,
    });
  }

  const total = roundMoney(
    items.reduce((sum, item) => sum + item.quantity * item.unit_price, 0)
  );

  const permitItems = items.filter((item) => item.requires_permit);
  if (permitItems.length > 0) {
    notes.push(`${permitItems.length} voci richiedono verifica titoli edilizi/permessi.`);
  }
  notes.push("Importi indicativi. Validare con preventivi d'impresa prima dell'impegno.");
  notes.push("Verificare agevolazioni fiscali applicabili con il commercialista.");

  return { items, total_estimated: total, notes };
}

function roundMoney(value: number): number {
  return Math.round(value * 100) / 100;
}
